allennlp/common/from_params.py

```diff
@@ -0,0 +1,185 @@
+from typing import TypeVar, Type, Dict, Union, Any, cast
+import inspect
+import logging
+
+from allennlp.common.checks import ConfigurationError
+from allennlp.common.params import Params
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+T = TypeVar('T')
+
+# getattr is necessary because mypy insists that no such attribute exists.
+_NO_DEFAULT = getattr(inspect, '_empty')  # pylint: disable=invalid-name
+
+def takes_arg(obj, arg: str) -> bool:
+    """
+    Checks whether the provided obj takes a certain arg.
+    If it's a class, we're really checking whether its constructor does.
+    If it's a function or method, we're checking the object itself.
+    Otherwise, we raise an error.
+    """
+    if inspect.isclass(obj):
+        signature = inspect.signature(obj.__init__)
+    elif inspect.ismethod(obj) or inspect.isfunction(obj):
+        signature = inspect.signature(obj)
+    else:
+        raise ConfigurationError(f"object {obj} is not callable")
+    return arg in signature.parameters
+
+def remove_optional(annotation):
+    """
+    Optional[X] annotations are actually represented as Union[X, NoneType].
+    For our purposes, the "Optional" part is not interesting, so here we
+    throw it away.
+    """
+    origin = getattr(annotation, '__origin__', None)
+    args = getattr(annotation, '__args__', ())
+    if origin == Union and len(args) == 2 and args[1] == type(None):
+        return args[0]
+    else:
+        return annotation
+
+def create_kwargs(cls: Type[T], params: Params, **extras) -> Dict[str, Any]:
+    """
+    Given some class, a `Params` object, and potentially other keyword arguments,
+    create a dict of keyword args suitable for passing to the class's constructor.
+    """
+    # Get the signature of the constructor.
+    signature = inspect.signature(cls.__init__)
+    kwargs: Dict[str, Any] = {}
+
+    # Iterate over all the constructor parameters and their annotations.
+    for name, param in signature.parameters.items():
+        # Skip "self". You're not *required* to call the first parameter "self",
+        # so in theory this logic is fragile, but if you don't call the self parameter
+        # "self" you kind of deserve what happens.
+        if name == "self":
```

Comment:
Seems like you can just skip the first parameter? The [parameters attribute is ordered](https://docs.python.org/3/library/inspect.html#inspect.Signature), so I think you can just skip the first one.

Reply:
Although in principle I agree that if you don't call the first parameter self you deserve it